src/scheduler/gc_work.rs

```diff
@@ -749,61 +810,25 @@ pub trait ScanObjectsWork<VM: VMBinding>: GCWork<VM> + Sized {
     /// the objects in this packet.
     fn make_another(&self, buffer: Vec<ObjectReference>) -> Self;
 
+    fn get_bucket(&self) -> WorkBucketStage;
+
     /// The common code for ScanObjects and PlanScanObjects.
     fn do_work_common(
         &self,
         buffer: &[ObjectReference],
         worker: &mut GCWorker<<Self::E as ProcessEdgesWork>::VM>,
-        mmtk: &'static MMTK<<Self::E as ProcessEdgesWork>::VM>,
+        _mmtk: &'static MMTK<<Self::E as ProcessEdgesWork>::VM>,
     ) {
         let tls = worker.tls;
+        assert!(!self.roots());
```

Comment:
I am not sure about performance implication here. But `debug_assert!` should be a safer option.